tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,455 @@
+      # not prepared, so we need to find them and do something later.
+      if current_node_name in previous_layers_visited:
+        nodes_from_other_layers.add(current_node_name)
+
+      elif current_node_name not in current_layer_visited:
+        subgraph.node.append(current_node)
+
+        current_layer_visited.add(current_node_name)
+        queue.extend(node_name_to_input_names[current_node_name])
+
+  return ExecutionSpec(
+      subgraph=subgraph,
+      input_names=_get_input_names_from_subgraph(subgraph),
+      output_names=set(output_names),
+      is_remote_op=False,
+      body_node_names=_get_body_node_names_from_subgraph(subgraph),
+      nodes_from_other_layers=nodes_from_other_layers)
+
+
+def _handle_nodes_from_other_layers(current_spec,
+                                    execution_specs,
+                                    graph,
+                                    node_name_to_node_def):
+  """Handle nodes that are from other layers."""
+  for node_name in current_spec.nodes_from_other_layers:
+    for previous_spec in execution_specs:
+
+      if node_name in previous_spec.body_node_names:
+        previous_spec.output_names.add(node_name)
+        current_spec.input_names.add(node_name)
+
+        node = node_name_to_node_def[node_name]
+        placeholder = _create_placeholder_node_from_existing_node(node, graph)
+        current_spec.subgraph.node.append(placeholder)
+
+  return current_spec
+
+
+def _partition_one_remote_op_layer(remote_op_names, node_name_to_input_names):
+  """Construct execution spec for remote ops."""
+  list_of_specs = []
+  for remote_op_name in remote_op_names:
+    spec = ExecutionSpec(
+        subgraph=None,
+        input_names=set(node_name_to_input_names[remote_op_name]),
+        output_names=set([remote_op_name]),
+        is_remote_op=True,
+        body_node_names=set([remote_op_name]),
+        nodes_from_other_layers=set([]))
+    list_of_specs.append(spec)
+
+  return list_of_specs
+
+
+def _create_placeholder_node(dtype, shape, name):
+  temp = tf.Graph()
+  with temp.as_default():
+    tf.compat.v1.placeholder(dtype=dtype, shape=shape, name=name)
+    return temp.as_graph_def().node[0]  # The first and the only node
+
+
+def _create_placeholder_node_from_existing_node(node, graph):
+  operation = graph.get_operation_by_name('import/%s' % (node.name))
+  dtype = operation.outputs[0].dtype
+  return _create_placeholder_node(dtype=dtype,
+                                  shape=None,
+                                  name=node.name)
+
+
+def _get_input_names_from_subgraph(subgraph):
+  input_names = {node.name for node in subgraph.node
+                 if _is_placeholder_op(node)}
+  return input_names
+
+
+def _get_body_node_names_from_subgraph(subgraph):
+  body_node_names = {node.name for node in subgraph.node
+                     if not _is_placeholder_op(node)}
+  return body_node_names
+
+
+class Relations:
```

Comment:
Optional: This could be expressed as a Python iterator / generator / generator function. (Ping me if you need a longer explanation for how to do this.)

Then you could write:
```
for remote_ops_one_layer in order:
```
instead of:
```
while not order.check_if_finished():
  remote_ops_one_layer = order.get_next_layer()
```

Reply:
Looking at some online tutorials

Reply:
Changed